todo_app/app.py

```diff
@@ -1,59 +1,138 @@
+    def load_user(user_id):
+        return User(user_id)
 
+    login_manager.init_app(app) 
+
+    @app.route('/login/callback')
+    def login():
+        #TODO - state check
+        
+        request_code = request.args.get('code')
+        client = WebApplicationClient(github_client_id)
+        
+        token_url, headers, body = client.prepare_token_request(
+            'https://github.com/login/oauth/access_token',
+            authorization_response=request.url,
+            code=request_code,
+            client_secret=github_client_secret
+        )
+
+        token_resp = requests.post(token_url, headers=headers, data=body)
+        client.parse_request_body_response(token_resp.text)
+
+        get_user_uri, headers, body = client.add_token('https://api.github.com/user')
+        user_info_resp = requests.get(get_user_uri, headers=headers, data=body)
+        user_id=user_info_resp.json()["id"]
+        user = User(user_id)
+        login_user(user)
+
+        return redirect('/')
 
     @app.route('/')
+    @login_required
     def index():
         todo_board = ToDoBoard(todo_board_id)
         items = todo_board.get_cards()
-        item_view_model = ViewModel(items)
+        item_view_model = ViewModel(items, current_user.role)
         return render_template('index.html', view_model=item_view_model)
 
     
     @app.route('/items', methods=['POST'])
+    @login_required
+    @writer_required
     def add_item():
         title = request.form.get('title')
         status = request.form.get('status')
         todo_board = ToDoBoard(todo_board_id)
         todo_board.create_card(title, status)
 
         items = todo_board.get_cards()
-        item_view_model = ViewModel(items)
+        item_view_model = ViewModel(items, current_user.role)
         return redirect('/')
 
     @app.route('/items/<item_id>')
+    @login_required
+    @writer_required
     def update_item_status(item_id):    
         status = request.args.get('status')
         todo_board = ToDoBoard(todo_board_id)
         todo_board.update_card_status(item_id, status)
 
         items = todo_board.get_cards()
-        item_view_model = ViewModel(items)
+        item_view_model = ViewModel(items, current_user.role)
         return render_template('index.html', view_model=item_view_model)
 
     if __name__ == '__main__':
         app.run()
 
     return app
 
+class User(UserMixin):
+    def __init__(self, id):
+        self.id = id
+        if id == 42806788:
+            self.role = "writer"
+        else:
+            self.role = "reader"
+
+    def get_id(self):
+        return self.id
+
+class MyCustomAnonymousUser(AnonymousUserMixin):
+    def __init__(self):
+        self.role = "writer"
+
+def writer_required(f):
+    @wraps(f)
+    def wrap(*args, **kwargs):
+        if current_user.role == "writer":
+            return f(*args, **kwargs)
+        else:
+            flash("Permission Denied: You do not have the required role to perform this action.")
```

Comment:
Nice; really clear UX if someone is trying to bypass our security/is being displayed the wrong options